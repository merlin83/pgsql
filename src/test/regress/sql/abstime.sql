-- **** testing built-in time types: abstime, reltime, and tinterval ****

--
-- timezones may vary based not only on location but the operating
-- system.  the main correctness issue is that the OS may not get 
-- daylight savings time right for times prior to Unix epoch (jan 1 1970).
--

CREATE TABLE ABSTIME_TBL (f1 abstime);

INSERT INTO ABSTIME_TBL (f1) VALUES ('Jan 14, 1973 03:14:21');

-- was INSERT INTO ABSTIME_TBL (f1) VALUES ('now'::abstime):
INSERT INTO ABSTIME_TBL (f1) VALUES ('Mon May  1 00:30:30 1995'::abstime);

INSERT INTO ABSTIME_TBL (f1) VALUES ('epoch'::abstime);

INSERT INTO ABSTIME_TBL (f1) VALUES ('current'::abstime);

INSERT INTO ABSTIME_TBL (f1) VALUES ('infinity'::abstime);

INSERT INTO ABSTIME_TBL (f1) VALUES ('-infinity'::abstime);

INSERT INTO ABSTIME_TBL (f1) VALUES ('May 10, 1943 23:59:12');


-- what happens if we specify slightly misformatted abstime? 
INSERT INTO ABSTIME_TBL (f1) VALUES ('Feb 35, 1946 10:00:00');

INSERT INTO ABSTIME_TBL (f1) VALUES ('Feb 28, 1984 25:08:10');


-- badly formatted abstimes:  these should result in invalid abstimes 
INSERT INTO ABSTIME_TBL (f1) VALUES ('bad date format');

INSERT INTO ABSTIME_TBL (f1) VALUES ('Jun 10, 1843');

-- test abstime operators

SELECT '' AS eight, ABSTIME_TBL.*;

SELECT '' AS six, ABSTIME_TBL.*
   WHERE ABSTIME_TBL.f1 < 'Jun 30, 2001'::abstime;

SELECT '' AS six, ABSTIME_TBL.*
   WHERE ABSTIME_TBL.f1 > '-infinity'::abstime;

SELECT '' AS six, ABSTIME_TBL.*
   WHERE 'May 10, 1943 23:59:12'::abstime <> ABSTIME_TBL.f1;

SELECT '' AS one, ABSTIME_TBL.*
   WHERE 'current'::abstime = ABSTIME_TBL.f1;

SELECT '' AS three, ABSTIME_TBL.*
   WHERE 'epoch'::abstime >= ABSTIME_TBL.f1;

SELECT '' AS four, ABSTIME_TBL.*
   WHERE ABSTIME_TBL.f1 <= 'Jan 14, 1973 03:14:21'::abstime;

SELECT '' AS four, ABSTIME_TBL.*
  WHERE ABSTIME_TBL.f1 <?>
	'["Apr 1 1945 00:00:00" "Dec 30 1999 23:00:00"]'::tinterval;

-- these four queries should return the same answer
-- the "infinity" and "-infinity" tuples in ABSTIME_TBL cannot be added and
-- therefore, should not show up in the results.
SELECT '' AS three, ABSTIME_TBL.*
  WHERE  (ABSTIME_TBL.f1 + '@ 3 year'::reltime) -- +3 years
	< 'Jan 14 14:00:00 1977'::abstime;

SELECT '' AS three, ABSTIME_TBL.*
   WHERE  (ABSTIME_TBL.f1 + '@ 3 year ago'::reltime)	-- -3 years
	< 'Jan 14 14:00:00 1971'::abstime;

SELECT '' AS three, ABSTIME_TBL.*
   WHERE  (ABSTIME_TBL.f1 - '@ 3 year'::reltime)        -- -(+3) years
	< 'Jan 14 14:00:00 1971'::abstime;

SELECT '' AS three, ABSTIME_TBL.*
   WHERE  (ABSTIME_TBL.f1 - '@ 3 year ago'::reltime)    -- -(-3) years
        < 'Jan 14 14:00:00 1977'::abstime;


SELECT '' AS ten, ABSTIME_TBL.*, RELTIME_TBL.*
   WHERE (ABSTIME_TBL.f1 + RELTIME_TBL.f1)
	< 'Jan 14 14:00:00 1971'::abstime;

