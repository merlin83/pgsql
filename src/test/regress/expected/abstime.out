QUERY: CREATE TABLE ABSTIME_TBL (f1 abstime);
QUERY: INSERT INTO ABSTIME_TBL (f1) VALUES ('Jan 14, 1973 03:14:21');
QUERY: INSERT INTO ABSTIME_TBL (f1) VALUES ('Mon May  1 00:30:30 1995'::abstime);
QUERY: INSERT INTO ABSTIME_TBL (f1) VALUES ('epoch'::abstime);
QUERY: INSERT INTO ABSTIME_TBL (f1) VALUES ('current'::abstime);
QUERY: INSERT INTO ABSTIME_TBL (f1) VALUES ('infinity'::abstime);
QUERY: INSERT INTO ABSTIME_TBL (f1) VALUES ('-infinity'::abstime);
QUERY: INSERT INTO ABSTIME_TBL (f1) VALUES ('May 10, 1943 23:59:12');
QUERY: INSERT INTO ABSTIME_TBL (f1) VALUES ('Feb 35, 1946 10:00:00');
WARN:Bad abstime external representation 'Feb 35, 1946 10:00:00'
QUERY: INSERT INTO ABSTIME_TBL (f1) VALUES ('Feb 28, 1984 25:08:10');
WARN:Bad abstime external representation 'Feb 28, 1984 25:08:10'
QUERY: INSERT INTO ABSTIME_TBL (f1) VALUES ('bad date format');
WARN:Bad abstime external representation 'bad date format'
QUERY: INSERT INTO ABSTIME_TBL (f1) VALUES ('Jun 10, 1843');
QUERY: SELECT '' AS eight, ABSTIME_TBL.*;
eight|f1                          
-----+----------------------------
     |Sun Jan 14 03:14:21 1973 PST
     |Mon May 01 00:30:30 1995 PDT
     |epoch                       
     |current                     
     |infinity                    
     |-infinity                   
     |Mon May 10 23:59:12 1943 PST
     |invalid                     
(8 rows)

QUERY: SELECT '' AS six, ABSTIME_TBL.*
   WHERE ABSTIME_TBL.f1 < 'Jun 30, 2001'::abstime;
six|f1                          
---+----------------------------
   |Sun Jan 14 03:14:21 1973 PST
   |Mon May 01 00:30:30 1995 PDT
   |epoch                       
   |current                     
   |-infinity                   
   |Mon May 10 23:59:12 1943 PST
(6 rows)

QUERY: SELECT '' AS six, ABSTIME_TBL.*
   WHERE ABSTIME_TBL.f1 > '-infinity'::abstime;
six|f1                          
---+----------------------------
   |Sun Jan 14 03:14:21 1973 PST
   |Mon May 01 00:30:30 1995 PDT
   |epoch                       
   |current                     
   |infinity                    
   |Mon May 10 23:59:12 1943 PST
(6 rows)

QUERY: SELECT '' AS six, ABSTIME_TBL.*
   WHERE 'May 10, 1943 23:59:12'::abstime <> ABSTIME_TBL.f1;
six|f1                          
---+----------------------------
   |Sun Jan 14 03:14:21 1973 PST
   |Mon May 01 00:30:30 1995 PDT
   |epoch                       
   |current                     
   |infinity                    
   |-infinity                   
(6 rows)

QUERY: SELECT '' AS one, ABSTIME_TBL.*
   WHERE 'current'::abstime = ABSTIME_TBL.f1;
one|f1     
---+-------
   |current
(1 row)

QUERY: SELECT '' AS three, ABSTIME_TBL.*
   WHERE 'epoch'::abstime >= ABSTIME_TBL.f1;
three|f1                          
-----+----------------------------
     |epoch                       
     |-infinity                   
     |Mon May 10 23:59:12 1943 PST
(3 rows)

QUERY: SELECT '' AS four, ABSTIME_TBL.*
   WHERE ABSTIME_TBL.f1 <= 'Jan 14, 1973 03:14:21'::abstime;
four|f1                          
----+----------------------------
    |Sun Jan 14 03:14:21 1973 PST
    |epoch                       
    |-infinity                   
    |Mon May 10 23:59:12 1943 PST
(4 rows)

QUERY: SELECT '' AS four, ABSTIME_TBL.*
  WHERE ABSTIME_TBL.f1 <?>
	'["Apr 1 1945 00:00:00" "Dec 30 1999 23:00:00"]'::tinterval;
four|f1                          
----+----------------------------
    |Sun Jan 14 03:14:21 1973 PST
    |Mon May 01 00:30:30 1995 PDT
    |epoch                       
    |current                     
(4 rows)

QUERY: SELECT '' AS three, ABSTIME_TBL.*
  WHERE  (ABSTIME_TBL.f1 + '@ 3 year'::reltime) 
	< 'Jan 14 14:00:00 1977'::abstime;
three|f1                          
-----+----------------------------
     |Sun Jan 14 03:14:21 1973 PST
     |epoch                       
     |Mon May 10 23:59:12 1943 PST
(3 rows)

QUERY: SELECT '' AS three, ABSTIME_TBL.*
   WHERE  (ABSTIME_TBL.f1 + '@ 3 year ago'::reltime)	
	< 'Jan 14 14:00:00 1971'::abstime;
three|f1                          
-----+----------------------------
     |Sun Jan 14 03:14:21 1973 PST
     |epoch                       
     |Mon May 10 23:59:12 1943 PST
(3 rows)

QUERY: SELECT '' AS three, ABSTIME_TBL.*
   WHERE  (ABSTIME_TBL.f1 - '@ 3 year'::reltime)        
	< 'Jan 14 14:00:00 1971'::abstime;
three|f1                          
-----+----------------------------
     |Sun Jan 14 03:14:21 1973 PST
     |epoch                       
     |Mon May 10 23:59:12 1943 PST
(3 rows)

QUERY: SELECT '' AS three, ABSTIME_TBL.*
   WHERE  (ABSTIME_TBL.f1 - '@ 3 year ago'::reltime)    
        < 'Jan 14 14:00:00 1977'::abstime;
three|f1                          
-----+----------------------------
     |Sun Jan 14 03:14:21 1973 PST
     |epoch                       
     |Mon May 10 23:59:12 1943 PST
(3 rows)

QUERY: SELECT '' AS ten, ABSTIME_TBL.*, RELTIME_TBL.*
   WHERE (ABSTIME_TBL.f1 + RELTIME_TBL.f1)
	< 'Jan 14 14:00:00 1971'::abstime;
ten|f1                          |f1              
---+----------------------------+----------------
   |epoch                       |@ 1 minute      
   |Mon May 10 23:59:12 1943 PST|@ 1 minute      
   |epoch                       |@ 5 hours       
   |Mon May 10 23:59:12 1943 PST|@ 5 hours       
   |epoch                       |@ 10 days       
   |Mon May 10 23:59:12 1943 PST|@ 10 days       
   |epoch                       |@ 3 months      
   |Mon May 10 23:59:12 1943 PST|@ 3 months      
   |epoch                       |@ 14 seconds ago
   |Mon May 10 23:59:12 1943 PST|@ 14 seconds ago
(10 rows)

